### PR TITLE
Add support for knocking to workers

### PR DESCRIPTION
**Pull request by dklimpel**
Fixes: #12806

Code was implemented with #6739, bot not used/activated.

Part of knocking only is:
 - [`^/_matrix/client/(api/v1|r0|v3|unstable)/knock/`](https://spec.matrix.org/v1.6/client-server-api/#post_matrixclientv3knockroomidoralias)

and not:
- `^/_matrix/client/(api/v1|r0|v3|unstable)/rooms/.*/(join|invite|leave|ban|unban|kick)$`

https://github.com/matrix-org/synapse/blob/8219525b66a2ffa9a9f1ed6e5c716fcc1146469e/synapse/rest/client/room.py#L928-L932

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Dirk Klimpel [5740567+dklimpel@users.noreply.github.com](mailto:5740567+dklimpel@users.noreply.github.com)

### CODE REVIEW

File: docker/configure_workers_and_start.py [hunk 1]
```diff
@@ -204,6 +204,7 @@
             "^/_matrix/client/(api/v1|r0|v3|unstable)/rooms/.*/send",
             "^/_matrix/client/(api/v1|r0|v3|unstable)/rooms/.*/(join|invite|leave|ban|unban|kick)$",
             "^/_matrix/client/(api/v1|r0|v3|unstable)/join/",
+            "^/_matrix/client/(api/v1|r0|v3|unstable)/knock/",
```
[the remark this answers is not in the document]
Comment (clokep) [reply docker/configure_workers_and_start.py:207]: (I'd ideally like to keep these in sync even if it doesn't FTR.)

[the remark this answers is not in the document]
Comment (dklimpel) [reply docker/configure_workers_and_start.py:207]: I have created https://github.com/matrix-org/sytest/pull/1332

**Comment by clokep** [comment]
@dklimpel Looks like all the knocking tests are failing when workers are used, so this sounds like it might *not* be safe after all?

**Comment by dklimpel** [comment]
Current main issue:

```
event_creator1 | 2023-02-22 11:55:37,717 - synapse.http.server - 123 - ERROR - POST-28 - Failed handle request via 'KnockRoomAliasServlet': <SynapseRequest at 0x7fb9bad4d490 method='POST' uri='/_matrix/client/v3/knock/%21jrbrudPMkOwghwHXsW:hs1?server_name=hs1' clientproto='HTTP/1.0' site='18013'>
  Traceback (most recent call last):
    File "/usr/local/lib/python3.11/site-packages/synapse/http/server.py", line 306, in _async_render_wrapper
      callback_return = await self._async_render(request)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.11/site-packages/synapse/http/server.py", line 516, in _async_render
      callback_return = await raw_callback_return
                        ^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.11/site-packages/synapse/rest/client/knock.py", line 84, in on_POST
      await self.room_member_handler.update_membership(
    File "/usr/local/lib/python3.11/site-packages/synapse/handlers/room_member.py", line 613, in update_membership
      result = await self.update_membership_locked(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.11/site-packages/synapse/handlers/room_member.py", line 1072, in update_membership_locked
      return await self.remote_knock(
             ^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.11/site-packages/synapse/handlers/room_member_worker.py", line 125, in remote_knock
      ret = await self._remote_knock_client(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.11/site-packages/synapse/replication/http/_base.py", line 237, in send_request
      data = await cls._serialize_payload(**kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by dklimpel** [comment]
current status

```
  event_creator1 | 2023-02-24 13:12:17,333 - synapse.http.server - 123 - ERROR - POST-9 - Failed handle request via 'KnockRoomAliasServlet': <SynapseRequest at 0x7f1f68b36490 method='POST' uri='/_matrix/client/v3/knock/%21xPTqglkCzOYcCiwtuD:hs1?server_name=hs1' clientproto='HTTP/1.0' site='18013'>
  Traceback (most recent call last):
    File "/usr/local/lib/python3.11/site-packages/synapse/http/server.py", line 306, in _async_render_wrapper
      callback_return = await self._async_render(request)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.11/site-packages/synapse/http/server.py", line 516, in _async_render
      callback_return = await raw_callback_return
                        ^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.11/site-packages/synapse/rest/client/knock.py", line 84, in on_POST
      await self.room_member_handler.update_membership(
    File "/usr/local/lib/python3.11/site-packages/synapse/handlers/room_member.py", line 614, in update_membership
      result = await self.update_membership_locked(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.11/site-packages/synapse/handlers/room_member.py", line 1073, in update_membership_locked
      return await self.remote_knock(
             ^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.11/site-packages/synapse/handlers/room_member_worker.py", line 126, in remote_knock
      ret = await self._remote_knock_client(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.11/site-packages/synapse/replication/http/_base.py", line 257, in send_request
      url_args = [
                 ^
    File "/usr/local/lib/python3.11/site-packages/synapse/replication/http/_base.py", line 258, in <listcomp>
      urllib.parse.quote(kwargs[name], safe="") for name in cls.PATH_ARGS
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.11/urllib/parse.py", line 867, in quote
      return quote_from_bytes(string, safe)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.11/urllib/parse.py", line 897, in quote_from_bytes
      raise TypeError("quote_from_bytes() expected bytes")
  TypeError: quote_from_bytes() expected bytes
```

**Comment by DMRobertson** [comment]
The arguments you pass to `_remote_knock_client` need to have the same names as the arguments to `_serialise_payload`

https://github.com/matrix-org/synapse/blob/9187fd940e2b2bbfd4df7204053cc26b2707aad4/synapse/replication/http/membership.py#L122-L142

and any arguments that appear in the URL instead of the request body need to have matching names in PATH_ARGS:

https://github.com/matrix-org/synapse/blob/9187fd940e2b2bbfd4df7204053cc26b2707aad4/synapse/replication/http/membership.py#L113

**Comment by DMRobertson** [comment]
Oh ignore me: I think that's already the case as of your PR. Hmm.

**Comment by dklimpel** [comment]
Thank your for the hints. It will help me.

**Comment by dklimpel** [comment]
Is there a hint how to fix ist?

```
2023-02-27T09:47:51.8325487Z ##[group][0;31m❌ TestKnocking/A_user_that_has_already_knocked_is_allowed_to_knock_again_on_the_same_room#01[0;37m (10ms)[0m
2023-02-27T09:47:51.8325973Z     knocking_test.go:359: MatchResponse got status 404 want 200 - http://127.0.0.1:33246/_matrix/client/v3/knock/%21TBuAgdkMCuKOqIesve:hs1?server_name=hs1 => {"errcode":"M_UNRECOGNIZED","error":"Unrecognized request"}
```

**Comment by DMRobertson** [comment]
> {"errcode":"M_UNRECOGNIZED","error":"Unrecognized request"}

I would guess that in the complement-with-workers config we're missing a listener/resource on some kind of worker.

There might be some clues in the top-level logs: https://github.com/matrix-org/synapse/actions/runs/4280704294/jobs/7452884787#step:6:5536

**Comment by dklimpel** [comment]
There is not test that call `ReplicationRemoteRescindKnockRestServlet` (`/_synapse/replication/remote_rescind_knock/:event_id`) at the moment.

**Review by clokep — APPROVED** [review]
Sorry for the delay here, thanks for fixing this! A real shame we had all the code and just never hooked it up. 😢

**Comment by clokep** [comment]
> There is not test that call `ReplicationRemoteRescindKnockRestServlet` (`/_synapse/replication/remote_rescind_knock/:event_id`) at the moment.

That's slightly concerning -- did you try it manually?

**Comment by dklimpel** [comment]
Unfortunately not.
I've only found that all tests pass, even if the servlet is commented out.

**Pull request merged**